jplag/src/main/java/de/jplag/clustering/algorithm/BFGS.java

```diff
@@ -0,0 +1,835 @@
+import org.apache.commons.math3.linear.DecompositionSolver;
+import org.apache.commons.math3.linear.LUDecomposition;
+import org.apache.commons.math3.linear.MatrixUtils;
+import org.apache.commons.math3.linear.RealMatrix;
+import org.apache.commons.math3.util.Pair;
+
+/**
+ * The Broyden–Fletcher–Goldfarb–Shanno (BFGS) algorithm is an iterative method for solving unconstrained nonlinear
+ * optimization problems.
+ * <p>
+ * The BFGS method belongs to quasi-Newton methods, a class of hill-climbing optimization techniques that seek a
+ * stationary point of a (preferably twice continuously differentiable) function. For such problems, a necessary
+ * condition for optimality is that the gradient be zero. Newton's method and the BFGS methods are not guaranteed to
+ * converge unless the function has a quadratic Taylor expansion near an optimum. However, BFGS has proven to have good
+ * performance even for non-smooth optimizations.
+ * <p>
+ * In quasi-Newton methods, the Hessian matrix of second derivatives is not computed. Instead, the Hessian matrix is
+ * approximated using updates specified by gradient evaluations (or approximate gradient evaluations). Quasi-Newton
+ * methods are generalizations of the secant method to find the root of the first derivative for multidimensional
+ * problems. In multi-dimensional problems, the secant equation does not specify a unique solution, and quasi-Newton
+ * methods differ in how they constrain the solution. The BFGS method is one of the most popular members of this class.
+ * <p>
+ * Like the original BFGS, the limited-memory BFGS (L-BFGS) uses an estimation to the inverse Hessian matrix to steer
+ * its search through variable space, but where BFGS stores a dense <code>n × n</code> approximation to the inverse
+ * Hessian (<code>n</code> being the number of variables in the problem), L-BFGS stores only a few vectors that
+ * represent the approximation implicitly. Due to its resulting linear memory requirement, the L-BFGS method is
+ * particularly well suited for optimization problems with a large number of variables (e.g., &gt; 1000). Instead of the
+ * inverse Hessian <code>H_k</code>, L-BFGS maintains * a history of the past <code>m</code> updates of the position
+ * <code>x</code> and gradient <code>∇f(x)</code>, where generally the history size <code>m</code> can be small (often
+ * <code>m &lt; 10</code>). These updates are used to implicitly do operations requiring the <code>H_k</code>-vector
+ * product.
+ * <h2>References</h2>
+ * <ol>
+ * <li>Roger Fletcher. Practical methods of optimization.</li>
+ * <li>D. Liu and J. Nocedal. On the limited memory BFGS method for large scale optimization. Mathematical Programming B
+ * 45 (1989) 503-528.</li>
+ * <li>Richard H. Byrd, Peihuang Lu, Jorge Nocedal and Ciyou Zhu. A limited memory algorithm for bound constrained
+ * optimization.</li>
+ * </ol>
+ * @author Haifeng Li
+ */
+public class BFGS {
```

Comment:
Not sure, but if this algorithm is simply copied, you could simply add the dependency

```
    <dependency>
      <groupId>com.github.haifengl</groupId>
      <artifactId>smile-core</artifactId>
      <version>2.6.0</version>
    </dependency>
```

Reply:
To be honest, the copied class is a bit problematic maintenance-wise. It would improve the quality of this PR a lot if we can replace it with a dependency. Could you please take a look at [this commit](https://github.com/dfuchss/jplag/commit/72d5b5357340f8be5bed6f9ee811020e32fe2609). I do not know what you changed in the copied code, so maybe you can tell us why you think it is not feasible. If we can replace the code with a dependency, we should!

Reply:
smile relies on BLAS and LAPACK for this, which is a pain to set up (at least I did not manage to). Apart from using different implementations of many functions I did the following:

```
505a168
>         int runs = 50;
521a185,186
>             } else if (runs-- < 0) {
>                 return f;
783c453
<         double dt = t[i];
---
>         double dt = i < n ? t[i] : 0;
```

First the fallback return, which is a dirty fix for an infinite loop which occasionally arises and I think the second fixes a bug in the original implementation.
At least without  these changes the spectral clustering fails sometimes.